airflow/utils/serialization.py

```diff
@@ -0,0 +1,290 @@
+        data = attr.asdict(o, recurse=True, filter=lambda a, v: a.init)  # type: ignore[arg-type]
+        dct[DATA] = serialize(data, depth + 1)
+        return dct
+
+    raise TypeError(f"cannot serialize object of type {cls}")
+
+
+def deserialize(o: T | None, full=True, type_hint: Any = None) -> object:
+    """
+    Deserializes an object of primitive type T into an object. Uses an allow
+    list to determine if a class can be loaded.
+
+    :param o: primitive to deserialize into an arbitrary object.
+    :param full: if False it will return a stringified representation
+                 of an object and will not load any classes
+    :param type_hint: if set it will be used to help determine what
+                 object to deserialize in. It does not override if another
+                 specification is found
+    :return: object
+    """
+    if o is None:
+        return o
+
+    if isinstance(o, _primitives):
+        return o
+
+    if isinstance(o, _iterables):
+        return [deserialize(d) for d in o]
+
+    if not isinstance(o, dict):
+        raise TypeError()
+
+    o = _convert(o)
+
+    # plain dict and no type hint
+    if CLASSNAME not in o and not type_hint:
+        return {str(k): deserialize(v, full) for k, v in o.items()}
+
+    # custom deserialization starts here
+    cls = None
+    version = 0
+    value = None
+    classname = None
+
+    if type_hint:
+        cls = type_hint
+        classname = qualname(cls)
+        version = 0  # type hinting always sets version to 0
+        value = o
+
+    if CLASSNAME in o and VERSION in o:
+        classname, version, value = decode(o)
+        if not _match(classname) and classname not in _extra_allowed:
+            raise ImportError(f"{classname} was not found in allow list for import")
+
+        if full:
+            cls = import_string(classname)
+
+    # only return string representation
+    if not full:
+        return _stringify(classname, version, value)
+
+    # registered deserializer
+    if classname in _deserializers:
+        return _deserializers[classname].deserialize(classname, version, value)
+
+    # class has deserialization function
+    if hasattr(cls, "deserialize"):
+        return getattr(cls, "deserialize")(value, version)
+
+    # attr or dataclass
+    if attr.has(cls) or dataclasses.is_dataclass(cls):
+        class_version = getattr(cls, "__version__", 0)
+        if int(version) > class_version:
+            raise TypeError(
+                "serialized version of %s is newer than module version (%s > %s)",
+                classname,
+                version,
+                class_version,
+            )
+
+        return cls(**deserialize(value))
+
+    # no deserializer available
+    raise TypeError(f"No deserializer found for {classname}")
+
+
+def _convert(old: dict) -> dict:
+    """Converts an old style serialization to new style"""
+    if OLD_TYPE in old and OLD_TYPE in old:
```

Comment:
same condition on either side of `and`?

Reply:
did you mean the following:

```suggestion
    if OLD_TYPE in old and OLD_DATA in old:
```

Reply:
Good catch!